pkg/controllers/machine/disruption/drift.go

```diff
@@ -28,81 +28,83 @@ import (
 
 	"github.com/aws/karpenter-core/pkg/apis/settings"
 	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/apis/v1beta1"
 	"github.com/aws/karpenter-core/pkg/cloudprovider"
 	"github.com/aws/karpenter-core/pkg/scheduling"
+	machineutil "github.com/aws/karpenter-core/pkg/utils/machine"
 )
 
-// Drift is a machine sub-controller that adds or removes status conditions on drifted machines
+// Drift is a nodeClaim sub-controller that adds or removes status conditions on drifted nodeClaims
 type Drift struct {
 	cloudProvider cloudprovider.CloudProvider
 }
 
-func (d *Drift) Reconcile(ctx context.Context, provisioner *v1alpha5.Provisioner, machine *v1alpha5.Machine) (reconcile.Result, error) {
-	hasDriftedCondition := machine.StatusConditions().GetCondition(v1alpha5.MachineDrifted) != nil
+func (d *Drift) Reconcile(ctx context.Context, nodePool *v1beta1.NodePool, nodeClaim *v1beta1.NodeClaim) (reconcile.Result, error) {
+	if !nodeClaim.StatusConditions().GetCondition(v1beta1.NodeLaunched).IsTrue() {
+		return reconcile.Result{}, nil
+	}
+
+	hasDriftedCondition := nodeClaim.StatusConditions().GetCondition(v1beta1.NodeDrifted) != nil
 
 	// From here there are three scenarios to handle:
-	// 1. If drift is not enabled but the machine is drifted, remove the status condition
+	// 1. If drift is not enabled but the nodeClaim is drifted, remove the status condition
 	if !settings.FromContext(ctx).DriftEnabled {
-		_ = machine.StatusConditions().ClearCondition(v1alpha5.MachineDrifted)
-		if hasDriftedCondition {
-			logging.FromContext(ctx).Debugf("removing drift status condition from machine as drift has been disabled")
-		}
-		return reconcile.Result{}, nil
-	}
-	// 2. If Machine is not launched, remove the drift status condition
-	if launchCond := machine.StatusConditions().GetCondition(v1alpha5.MachineLaunched); launchCond == nil || launchCond.IsFalse() {
-		_ = machine.StatusConditions().ClearCondition(v1alpha5.MachineDrifted)
+		_ = nodeClaim.StatusConditions().ClearCondition(v1beta1.NodeDrifted)
 		if hasDriftedCondition {
-			logging.FromContext(ctx).Debugf("removing drift status condition from machine as machine isn't launched")
+			logging.FromContext(ctx).Debugf("removing drift status condition, drift has been disabled")
```

Comment:
Don't we shortcircuit if drift is disabled? How can we get here?

Reply:
We do not short-circuit in this case. Drift being disabled causes us to remove the conditions. I believe this is the correct handling here.